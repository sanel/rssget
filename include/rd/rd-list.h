#ifndef __RD_LIST_H__
#define __RD_LIST_H__

#include <stdlib.h>

typedef struct _rd_list_node_t {
	void *value;
	struct _rd_list_node_t *next;
	struct _rd_list_node_t *prev;
} rd_list_node_t;

typedef struct _rd_list_t {
	unsigned int count;
	rd_list_node_t *tail;
} rd_list_t;

rd_list_t *
rd_list_init(void);

void
rd_list_fini(rd_list_t *o);

void
rd_list_clear(rd_list_t *o);

rd_list_node_t *
rd_list_insert(rd_list_t *lst, rd_list_node_t *it, void *value);

rd_list_node_t *
rd_list_erase(rd_list_t *lst, rd_list_node_t *it);

inline rd_list_node_t *
rd_list_first(rd_list_t *o) { return o->tail ? o->tail->next : NULL; }

/*
 * NOTE: it will return last element in list, which is tail. It is not last
 * actual element - you should obtain it with rd_list_last(o)->prev;
 */
inline rd_list_node_t *
rd_list_last(rd_list_t *o) { return o->tail ? o->tail : NULL; }

inline rd_list_node_t *
rd_list_rfirst(rd_list_t *o) { return o->tail ? o->tail->prev: NULL; }

inline rd_list_node_t *
rd_list_push_back(rd_list_t *o, void *value) {
	return rd_list_insert(o, rd_list_last(o), value);
}

inline rd_list_node_t *
rd_list_push_front(rd_list_t *o, void *value) {
	return rd_list_insert(o, rd_list_first(o), value);
}

inline unsigned int
rd_list_count(rd_list_t *o) { return o->count; }

/* forward/backward iterator emulation */
#define rd_list_next(it) (it)->next
#define rd_list_prev(it) (it)->prev
#define rd_list_rnext(it) (it)->prev
#define rd_list_rprev(it) (it)->next

#endif
