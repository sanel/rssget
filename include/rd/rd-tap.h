#ifndef __RD_TAP_H__
#define __RD_TAP_H__

/* TAP - test anything protocol */

void rd_plan(int n);
int  rd_done(void);

void rd_todo(const char *msg);
void rd_todo_end();

void rd_skip(int n, const char *msg);
void rd_skip_end();

int _rd_ok(int expr, const char *msg, const char *path, unsigned int line);
int _rd_is_str(const char *s1, const char *s2, const char *msg, const char *path, unsigned int line);

#define rd_ok(expr, str)        _rd_ok(expr, str, __FILE__, __LINE__)
#define rd_not_ok(expr, str)    _rd_ok(!(expr), str, __FILE__, __LINE__)
#define rd_is_str(s1, s2, str)  _rd_is_str(s1, s2, str, __FILE__, __LINE__)

#endif
