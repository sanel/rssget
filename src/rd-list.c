#include <rd/rd-utils.h>
#include <rd/rd-list.h>
#include <stdio.h>

rd_list_t *
rd_list_init(void) {
	rd_list_t *ret = malloc(sizeof(rd_list_t));
	RD_RETURN_VAL_IF_FAIL(ret != NULL, NULL);

	ret->count = 0;
	ret->tail  = NULL;
	return ret;
}

void
rd_list_fini(rd_list_t *o) {
	RD_RETURN_IF_FAIL(o != NULL);
	rd_list_clear(o);
	free(o);
}

void
rd_list_clear(rd_list_t *o) {
	RD_RETURN_IF_FAIL(o != NULL);
	rd_list_node_t *tail, *p, *t;

	if (RD_UNLIKELY(rd_list_count(o) == 0))
		return;

	tail = o->tail;
	p = tail->next;
	while (p != tail) {
		t = p->next;
		if (p->value) free(p->value);
		free(p);
		p = t;
	}

	free(o->tail);
	o->tail = NULL;
	o->count = 0;
}

rd_list_node_t *
rd_list_insert(rd_list_t *lst, rd_list_node_t *it, void *value) {
	rd_list_node_t *tmp = malloc(sizeof(rd_list_node_t));
	tmp->value = value;
	tmp->next = tmp->prev = 0;

	if (!lst->tail) {
		/* dummy node if not present */
		rd_list_node_t *t = malloc(sizeof(rd_list_node_t));
		RD_ASSERT(it == NULL);

		t->value = NULL;
		t->next = t->prev = tmp;
		tmp->next = tmp->prev = t;
		lst->tail = t;
	} else {
		tmp->next = it;
		tmp->prev = it->prev;
		if (it->prev) it->prev->next = tmp;
		it->prev = tmp;
	}

	lst->count++;
	return tmp;
}

rd_list_node_t *
rd_list_erase(rd_list_t *lst, rd_list_node_t *it) {
	RD_ASSERT(it != lst->tail && "Bad code! You are erasing tail");
	rd_list_node_t *ret;

	it->prev->next = it->next;
	it->next->prev = it->prev;
	ret = it->next;

	free(it->value);
	free(it);
	lst->count--;

	/* returns next element */
	return ret;
}

