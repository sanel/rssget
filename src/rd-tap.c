#include <rd/rd-tap.h>
#include <stdio.h>
#include <string.h>

static int texpected;
static int tfailed;
static int tcurr;
static const char *todo_msg;

void rd_plan(int n) {
	texpected = n;
	printf("1..%i\n", n);
}

int rd_done(void) {
	int ret = 0;

	if (texpected > tcurr)  {
		printf("# Looks like you have planned %i but executed %i tests\n", texpected, tcurr);
		ret = 1;
	}

	if (tfailed) {
		printf("# Looks like you have failed %i tests out of %i run and %i planned\n",
			   tfailed, tcurr, texpected);
		ret = 1;
	}

	return ret;
}

int _rd_ok(int expr, const char *name, const char *path, unsigned int line) {
	if (!expr) printf("not ");
	printf("ok %i", ++tcurr);
	if (name) printf(" - %s", name);
	if (todo_msg) printf(" # TODO %s", todo_msg);

	printf("\n");

	if (!expr) {
		printf("#   Failed");
		if (todo_msg) printf("(TODO) ");
		printf("test ");
		if (name) printf("'%s'\n#   ", name);
		printf("at %s line %i.\n", path, line);
		if(!todo_msg)
			tfailed++;
	}

	return expr;
}

int _rd_is_str(const char *s1, const char *s2, const char *msg, const char *path, unsigned int line) {
	return _rd_ok(strcmp(s1, s2) == 0, msg, path, line);
}

void rd_todo(const char *msg) {
	todo_msg = msg;
}

void rd_todo_end(void) {
	todo_msg = 0;
}
