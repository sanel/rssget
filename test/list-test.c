#include <rd/rd-list.h>
#include <rd/rd-tap.h>
#include <string.h>

static int* alloc_num(int n) {
	int *r = malloc(sizeof(int));
	*r = n;
	return r;
}

static int as_num(void *n) {
	return *(int*)n;
}

int main() {
	rd_list_t *lst;
	rd_list_node_t *it;
	int i;

	rd_plan(14);

	lst = rd_list_init();
	rd_ok(rd_list_count(lst) == 0, "empty list");
	rd_ok(rd_list_first(lst) == NULL, "no first element");
	rd_ok(rd_list_last(lst) == NULL, "no last element");

	for(i = 0; i < 10; i++)
		rd_list_push_back(lst, alloc_num(i));

	rd_ok(rd_list_count(lst) == 10, "filled list");

	for(it = rd_list_first(lst), i = 0; it != rd_list_last(lst); it = it->next, i++)
		rd_ok(i == as_num(it->value), "compare elements in list");

	rd_list_clear(lst);

	rd_ok(rd_list_count(lst) == 0, "empty list");
	rd_ok(rd_list_first(lst) == NULL, "no first element");
	rd_ok(rd_list_last(lst) == NULL, "no last element");

	/* clear it multiple times */
	rd_list_clear(lst);
	rd_list_clear(lst);

	rd_ok(rd_list_count(lst) == 0, "empty list");
	rd_ok(rd_list_first(lst) == NULL, "no first element");
	rd_ok(rd_list_last(lst) == NULL, "no last element");

	/* manual insert; things are inserted in reverse order */
	it = rd_list_insert(lst, 0, alloc_num(10));
	it = rd_list_insert(lst, it, alloc_num(20));
	it = rd_list_insert(lst, it, alloc_num(30));

	it = rd_list_first(lst);
	rd_ok(30 == as_num(it->value), "first element in manual insert");

	it = it->next;
	rd_ok(20 == as_num(it->value), "second element in manual insert");

	it = it->next;
	rd_ok(10 == as_num(it->value), "third element in manual insert");
	rd_ok(rd_list_count(lst) == 3, "number of elements after manual insert");

	/* erase middle element */
	it = it->prev;
	it = rd_list_erase(lst, it);
	rd_ok(10 == as_num(it->value), "erasing returns iterator to next element");
	rd_ok(rd_list_count(lst) == 2, "number of elements after erasing middle element");

	rd_list_clear(lst);

	/* reverse iterating */
	rd_list_push_back(lst, alloc_num(10));
	rd_list_push_back(lst, alloc_num(20));
	rd_list_push_back(lst, alloc_num(30));

	it = rd_list_rfirst(lst);
	rd_ok(30 == as_num(it->value), "first element in reverse order");
	it = rd_list_rnext(it);
	rd_ok(20 == as_num(it->value), "second element in reverse order");
	it = rd_list_rnext(it);
	rd_ok(10 == as_num(it->value), "third element in reverse order");

	rd_list_fini(lst);
}
