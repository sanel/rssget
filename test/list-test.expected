1..14
ok 1 - empty list
ok 2 - no first element
ok 3 - no last element
ok 4 - filled list
ok 5 - compare elements in list
ok 6 - compare elements in list
ok 7 - compare elements in list
ok 8 - compare elements in list
ok 9 - compare elements in list
ok 10 - compare elements in list
ok 11 - compare elements in list
ok 12 - compare elements in list
ok 13 - compare elements in list
ok 14 - compare elements in list
ok 15 - empty list
ok 16 - no first element
ok 17 - no last element
ok 18 - empty list
ok 19 - no first element
ok 20 - no last element
ok 21 - first element in manual insert
ok 22 - second element in manual insert
ok 23 - third element in manual insert
ok 24 - number of elements after manual insert
ok 25 - erasing returns iterator to next element
ok 26 - number of elements after erasing middle element
ok 27 - first element in reverse order
ok 28 - second element in reverse order
ok 29 - third element in reverse order
